okhttp/src/jvmMain/kotlin/okhttp3/internal/-HttpUrlJvm.kt

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (C) 2022 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal
+
+import java.nio.charset.Charset
+import okhttp3.internal.CommonHttpUrl.FORM_ENCODE_SET
+import okhttp3.internal.CommonHttpUrl.HEX_DIGITS
+import okhttp3.internal.CommonHttpUrl.isPercentEncoded
+import okhttp3.internal.JvmHttpUrl.canonicalizeWithCharset
+import okio.Buffer
+
+internal object JvmHttpUrl {
+
+  internal fun Buffer.writeCanonicalized(
+    input: String,
+    pos: Int,
+    limit: Int,
+    encodeSet: String,
+    alreadyEncoded: Boolean,
+    strict: Boolean,
+    plusIsSpace: Boolean,
+    unicodeAllowed: Boolean,
+    charset: Charset?
+  ) {
+    var encodedCharBuffer: Buffer? = null // Lazily allocated.
+    var codePoint: Int
+    var i = pos
+    while (i < limit) {
+      codePoint = input.codePointAt(i)
+      if (alreadyEncoded && (codePoint == '\t'.code || codePoint == '\n'.code ||
+          codePoint == '\u000c'.code || codePoint == '\r'.code)) {
+        // Skip this character.
+      } else if (codePoint == ' '.code && encodeSet === FORM_ENCODE_SET) {
+        // Encode ' ' as '+'.
+        writeUtf8("+")
+      } else if (codePoint == '+'.code && plusIsSpace) {
+        // Encode '+' as '%2B' since we permit ' ' to be encoded as either '+' or '%20'.
+        writeUtf8(if (alreadyEncoded) "+" else "%2B")
+      } else if (codePoint < 0x20 ||
+        codePoint == 0x7f ||
+        codePoint >= 0x80 && !unicodeAllowed ||
+        codePoint.toChar() in encodeSet ||
+        codePoint == '%'.code &&
+        (!alreadyEncoded || strict && !input.isPercentEncoded(i, limit))) {
+        // Percent encode this character.
+        if (encodedCharBuffer == null) {
+          encodedCharBuffer = Buffer()
+        }
+
+        if (charset == null || charset == Charsets.UTF_8) {
+          encodedCharBuffer.writeUtf8CodePoint(codePoint)
+        } else {
+          encodedCharBuffer.writeString(input, i, i + Character.charCount(codePoint), charset)
+        }
+
+        while (!encodedCharBuffer.exhausted()) {
+          val b = encodedCharBuffer.readByte().toInt() and 0xff
+          writeByte('%'.code)
+          writeByte(HEX_DIGITS[b shr 4 and 0xf].code)
+          writeByte(HEX_DIGITS[b and 0xf].code)
+        }
+      } else {
+        // This character doesn't need encoding. Just copy it over.
+        writeUtf8CodePoint(codePoint)
+      }
+      i += Character.charCount(codePoint)
```

Comment:
Is this the reason it isn’t multipatform? Okio _almost_ does this: 
https://square.github.io/okio/3.x/okio/okio/okio/utf8-size.html

Reply:
I’d prefer we create our own multiplatform implementation of `Character.charCount()` vs. duplicate this whole function?

Reply:
. . . or is the issue the Charset parameter?

It’d be interesting for OkHttp to define a `Charset` that only permits UTF-8 on non-JVM platforms!

Reply:
Should this go down to okio?